### PR TITLE
Dynamic ref pad

### PR DESCRIPTION
**Pull request by CharlieL7**
* extends `pad` operator to handle dynamic input shapes
* only handles computing the shape for adding constant padding to a dynamic shape
  * adds the padding to the min, max, and opt values (unless opt is 0, where it keeps it 0)
  * does not handle `reflect` padding with dynamic shapes
* I don't think the `pad_mode` is used, may want to refactor later

### CODE REVIEW

File: src/include/migraphx/shape.hpp
Comment (TedThemistokleous) [inline src/include/migraphx/shape.hpp:114]: No subtract dynamic_dimension from size_t?  Are we assuming this result is just a size_t then and handled somewhere else?

Comment (CharlieL7) [reply src/include/migraphx/shape.hpp:114]: I don't use that case in the current code. I will add it and some asserts for the subtraction operators to catch if the subtraction goes below 0. (a dimension of 0 won't make sense for most cases either?)

Comment (CharlieL7) [reply src/include/migraphx/shape.hpp:114]: It's a somewhat weird case however.
Example:
10 - {2, 5, 3} = {5, 8, 7} where the `min` output is `10 - max` and `max` output is `10 - min`

Comment (CharlieL7) [reply src/include/migraphx/shape.hpp:114]: Holding off on adding that case. I don't see where we would use it.

Comment (TedThemistokleous) [reply src/include/migraphx/shape.hpp:114]: Thats fine. Just wanted to ask the question since you had the other cases with the addition case.

**Review by TedThemistokleous — APPROVED** [review]
Looks good.  Should be good to go in since you added asserts if we get into those cases i mentioned above.

File: src/include/migraphx/op/pad.hpp
Comment (umangyadav) [inline src/include/migraphx/op/pad.hpp:69]: We can also add `+=` friend method.

**Comment by umangyadav** [comment]
>I don't think the pad_mode is used, may want to refactor later 
Can you open an issue to track?

File: src/shape.cpp
Comment (CharlieL7) [reply src/shape.cpp:550]: `opt` is implied to have to be within the range of `min` and `max`, but we currently don't prevent it being otherwise. The assert should catch if we're going to wraparound on `size_t`.

File: src/include/migraphx/op/pad.hpp
Comment (CharlieL7) [reply src/include/migraphx/op/pad.hpp:69]: `+=` function is mutating so I don't think it would benefit from being implemented outside of the class declaration. Adding the operators as class functions.

File: src/shape.cpp
Comment (umangyadav) [inline src/shape.cpp]: Should this by `x` instead of `y` ?

Comment (CharlieL7) [reply src/shape.cpp]: Should've just fixed that typo in commit `54ccaea`.

**Review by umangyadav — APPROVED** [review]

File: src/shape.cpp
Comment (pfultz2) [inline src/shape.cpp:511]: Use an `if` statement: `if (this->opt != 0) this->opt += x`.

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**
